### PR TITLE
[ir] [refactor] Add a `verify` pass to find out illegal IRs, and remove OffloadedStmt::begin_stmt/end_stmt

### PR DESCRIPTION
**Pull request by xumingkuan**
Related issue = https://github.com/taichi-dev/taichi/pull/727#issuecomment-611296391

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: taichi/analysis/verify.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
+#include "taichi/ir/ir.h"
+#include <vector>
+#include <unordered_set>
+
+TLANG_NAMESPACE_BEGIN
+
+class IRVerifier : public IRVisitor {
```
Comment (archibate) [inline taichi/analysis/verify.cpp:7]: Maybe we want `BasicStmtVisitor`? So that stmts inside `IfStmt` or `WhileStmt` can be verified.

Comment (xumingkuan) [reply taichi/analysis/verify.cpp:7]: We want to add more checks involving operands, so I'll override visitor functions for all container statements.

**Comment by xumingkuan** [comment]
We already have some tests (test_sparse_parallel.py for example) with an `OffLoadedStmt` whose `parent` is `nullptr`...

File: taichi/analysis/verify.cpp (hunk 1 above)
Comment (xumingkuan) [reply taichi/analysis/verify.cpp:7]: Oh I realized that they are implemented in the visitor of `Block`. So we can use `BasicStmtVisitor` directly.

File: taichi/analysis/verify.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    current_block = stmt_list;
+    visible_stmts.emplace_back();
+    for (auto &stmt : stmt_list->statements) {
+      stmt->accept(this);
+    }
+    current_block = backup_block;
+    visible_stmts.pop_back();
+  }
+
+  void visit(LocalLoadStmt *stmt) override {
```
Comment (yuanming-hu) [inline taichi/analysis/verify.cpp:56]: We can do the same test for `LocalStoreStmt` :-)

**Comment by xumingkuan** [comment]
Tests are still failing if https://github.com/taichi-dev/taichi/pull/731/commits/4bc02a13cdc59952e9129f7b4124a43d6735167e is reverted.

**Comment by xumingkuan** [comment]
25 tests are still failing if we do this `verify` after each `typecheck(ir)`.

File: taichi/transforms/type_check.cpp [hunk 3]
```diff
@@ -412,7 +412,11 @@ namespace irpass {
 
 void typecheck(IRNode *root) {
   check_fields_registered(root);
-  return TypeCheck::run(root);
+  TypeCheck::run(root);
+  //  if (root->is<Block>() && root->as<Block>()->parent == nullptr) {
```
Comment (yuanming-hu) [inline taichi/transforms/type_check.cpp:416]: Are these comments necessary here?

Comment (xumingkuan) [reply taichi/transforms/type_check.cpp:416]: I think we can `verify` after each `typecheck`. However, 23 tests are failing if I uncomment this now...

Comment (yuanming-hu) [reply taichi/transforms/type_check.cpp:416]: Sounds good - I'll merge this in once CI passes. Let's fix other issues later in separate PRs. Thanks so much!!

File: taichi/transforms/fix_block_parents.cpp [hunk 4]
```diff
@@ -46,6 +46,7 @@ class FixBlockParents : public IRVisitor {
   }
 
   void visit(OffloadedStmt *stmt) override {
+    stmt->parent = current_block; // hot fix
```
Comment (xumingkuan) [inline taichi/transforms/fix_block_parents.cpp:49]: Some tests are failing without this hotfix. This pass should be fixing `Block`'s parents but not `OffloadedStmt`'s. I think we can fix this systematically in the future.

File: taichi/transforms/compile_to_offloads.cpp [hunk 5]
```diff
@@ -90,6 +90,8 @@ void compile_to_offloads(IRNode *ir,
 
   // Final field registration correctness & type checking
   irpass::typecheck(ir);
+  irpass::fix_block_parents(ir);  // hot fix
```
Comment (xumingkuan) [inline taichi/transforms/compile_to_offloads.cpp:93]: And this is actually only for fixing `OffloadedStmt`'s parents.

Comment (yuanming-hu) [reply taichi/transforms/compile_to_offloads.cpp:93]: All sounds good. Let's merge this in and deal with the rest later. Thanks again.

**Pull request merged**